src/core/qgsshapegenerator.cpp

```diff
@@ -19,43 +19,54 @@
 #include "qgsgeometryutils.h"
 #include <QLineF>
 #include <QList>
+#include <QPainterPath>
 #include <algorithm>
 
-QLineF segment( int index, QRectF rect )
+QLineF segment( int index, QRectF rect, double radius )
 {
+  const int yMultiplier = rect.height() < 0 ? -1 : 1;
   switch ( index )
   {
     case 0:
-      return QLineF( rect.left(),
+      return QLineF( rect.left() + radius,
                      rect.top(),
-                     rect.right(),
+                     rect.right() - radius,
                      rect.top() );
     case 1:
       return QLineF( rect.right(),
-                     rect.top(),
+                     rect.top() + yMultiplier * radius,
                      rect.right(),
-                     rect.bottom() );
+                     rect.bottom() - yMultiplier * radius );
     case 2:
-      return QLineF( rect.right(),
+      return QLineF( rect.right() - radius,
                      rect.bottom(),
-                     rect.left(),
+                     rect.left() + radius,
                      rect.bottom() );
     case 3:
       return QLineF( rect.left(),
-                     rect.bottom(),
+                     rect.bottom() - yMultiplier * radius,
                      rect.left(),
-                     rect.top() );
+                     rect.top() + yMultiplier * radius );
     default:
       return QLineF();
   }
 }
 
 QPolygonF QgsShapeGenerator::createBalloon( const QgsPointXY &origin, const QRectF &rect, double wedgeWidth )
```

Comment:
Any reason why you're not unlocking rounded corners here?

Reply:
I'll eventually remove that method entirely in favour of the painter path version only